src/cursor.rs

```diff
@@ -0,0 +1,167 @@
+use std::{
+    collections::VecDeque,
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+    time::Duration,
+};
+
+use bson::{doc, Document};
+use derivative::Derivative;
+use futures::{future::BoxFuture, stream::Stream};
+
+use crate::{
+    error::Result,
+    operation::GetMore,
+    options::StreamAddress,
+    results::GetMoreResult,
+    Client,
+    Namespace,
+    RUNTIME,
+};
+
+#[derive(Debug)]
+pub(crate) struct CursorSpecification {
+    pub(crate) ns: Namespace,
+    pub(crate) address: StreamAddress,
+    pub(crate) id: i64,
+    pub(crate) batch_size: Option<u32>,
+    pub(crate) max_time: Option<Duration>,
+    pub(crate) buffer: VecDeque<Document>,
+}
+
+#[derive(Debug)]
+pub struct Cursor {
+    client: Client,
+    get_more: GetMore,
+    exhausted: bool,
+    state: State,
+}
+
+type GetMoreFuture = BoxFuture<'static, Result<GetMoreResult>>;
+
+/// Describes the current state of the Cursor. If the state is Executing, then a getMore operation
+/// is in progress. If the state is Buffer, then there are documents available from the current
+/// batch.
+#[derive(Derivative)]
+#[derivative(Debug)]
+enum State {
+    Executing(#[derivative(Debug = "ignore")] GetMoreFuture),
+    Buffer(VecDeque<Document>),
+    Exhausted,
+}
+
+impl Cursor {
+    pub(crate) fn new(client: Client, spec: CursorSpecification) -> Self {
+        let get_more = GetMore::new(
+            spec.ns,
+            spec.id,
+            spec.address,
+            spec.batch_size,
+            spec.max_time,
+        );
+
+        Self {
+            client,
+            get_more,
+            exhausted: spec.id == 0,
+            state: State::Buffer(spec.buffer),
+        }
+    }
+}
+
+impl Drop for Cursor {
+    fn drop(&mut self) {
+        if self.exhausted {
+            return;
+        }
+
+        let namespace = self.get_more.namespace().clone();
+        let client = self.client.clone();
+        let cursor_id = self.get_more.cursor_id();
+
+        RUNTIME.execute(async move {
+            let _: Result<_> = client
+                .database(&namespace.db)
+                .run_command(
+                    doc! {
+                        "killCursors": &namespace.coll,
+                        "cursors": [cursor_id]
+                    },
+                    None,
+                )
+                .await;
+        })
+    }
+}
+
+impl Stream for Cursor {
+    type Item = Result<Document>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+        loop {
+            let exhausted = self.exhausted;
```

Comment:
is this copy necessary, or can we just use `self.exhausted`?

Reply:
Using `self.exhausted` requires mutably borrowing `self` again, which we can't do since we're borrowing `self.state`. Booleans are about as small as things get though, so I don't think it's much of issue.

Reply:
the `Executing` branch of the match statement uses `self.exhausted` though, which is why I was wondering. I just tried it out and it seems to compile using `self.exhausted` on line 144 which was previously the only usage of the copy. I'm not really concerned about the performance from copying it, just more about eliminating extra variables if we don't need them.

Reply:
Interesting. At some point during my debugging, it was necessary, but I guess it's not anymore. I've updated it to get rid of that variable.